Mostly GitHub workflow and build system changes

jeremy-murphy opened the pull request:
And some fixes to the documentation, and one fix to the code.

Also updated the maintainer to myself, which I should have done a few years ago when I took over Boost.Graph, whoops.

Commit:
Remove mentions of bsp_process_group from docs

J5lx committed:
Document that metis_reader expects a seekable stream

Lastique committed:
Updated Boost.Filesystem usage.

Removed usage of Boost.Filesystem APIs that were deprecated and
then removed.

jeremy-murphy committed:
Merge pull request #35 from Lastique/feature/update_filesystem_usage

Update Boost.Filesystem usage

grafikrobot committed:
Make the library modular usable.

grafikrobot committed:
Switch to library requirements instead of source. As source puts extra source in install targets.

grafikrobot committed:
Add missing NO_LIB usage requirements.

grafikrobot committed:
Add requires-b2 check to top-level build file.

grafikrobot committed:
Bump B2 require to 5.2

grafikrobot committed:
Move inter-lib dependencies to a project variable and into the build targets.

grafikrobot committed:
Add python build setup.

grafikrobot committed:
Use latest ubuntu/compilers on GHA.

grafikrobot committed:
Move include to target.

pdimov committed:
Merge pull request #38 from grafikrobot/modular

Add support for modular build structure.

jeremy-murphy committed:
Merge pull request #34 from J5lx/metis-reader-seeking

Document that metis_reader expects a seekable stream

jeremy-murphy committed:
Merge pull request #24 from cojuer/develop

Remove mentions of bsp_process_group from docs

jeremy-murphy committed:
Update maintainer

Should have done this when I took over Boost.Graph, whoops.

jzmaddock commented:
The changes look fine to me.  Any idea what's started causing all the segfaults in the tests?

jzmaddock commented:
BTW we could really use an MSVC CI runner too.

jeremy-murphy committed:
Remove C++03,11,2a from GHA

pdimov committed:
Try libopenmpi-dev on 24.04 as well

aminiussi commented:
I test boost.mpi with HPE/Cray MPI which is mpich based so nothing trivial comes to mind. 

I only test with cmake and I think there are no cmake based tests for graph parallel, so I'll have to make a b2/gcc13/mpich build on my ubuntu laptop to sees if I can reproduce the issue.

aminiussi commented:
I would replace:
```
mpi_process_group pg;
```
with
```
boost::mpi::communicator world{};
BOOST_TEST(world.size() > 1); // not supposed to work on one proc
mpi_process_group pg{world};
```
Then, I suspect the test will fail the `BOOST_TEST(world.size() > 1)`  check as the communicator size is  1 (and that is the real problem).

Since the mpi_process_group class allocate one incomming and outgoing communication structures in containers of size comm.size(), sending a message to proc 1 will retrieve a non existing outgoing structure   (in mpi_process_group.ipp:173).

I suspect the problem is in the project_config.jam generated file that does not actually select the mpich implementation, which ends up launching 2 singleton instead of two process MPI jobs.

Lastique committed:
Remove dependencies on Boost.StaticAssert.

Boost.StaticAssert has been merged into Boost.Config, so remove
the dependency.

pdimov committed:
Merge pull request #45 from Lastique/feature/remove_static_assert

Remove dependencies on Boost.StaticAssert
